File: python-web/lib/confluence_client.py
"""Confluence Cloud REST API client.

Uses basic auth (email + API token) to fetch release note pages,
their content in storage format (XHTML), and image attachments.
"""

import time
import logging
from pathlib import Path

import requests

logger = logging.getLogger(__name__)


class ConfluenceClient:
    """Client for Confluence Cloud REST API."""

    def __init__(self, base_url: str, email: str, api_token: str):
        self.base_url = base_url.rstrip("/")
        self.session = requests.Session()
        self.session.auth = (email, api_token)
        self.session.headers.update({"Accept": "application/json"})

    # ── Core request with retry ──────────────────────────────────────

    def _get(self, url: str, params: dict | None = None) -> dict:
        """GET request with exponential backoff on 429."""
        for attempt in range(5):
            resp = self.session.get(url, params=params)
            if resp.status_code == 429:
                wait = 2 ** attempt
                logger.warning("Rate limited (429), waiting %ds...", wait)
                time.sleep(wait)
                continue
            resp.raise_for_status()
            return resp.json()
        raise RuntimeError("Rate limit exceeded after 5 retries")

    # ── Page operations (REST API v2) ────────────────────────────────

    def get_child_pages(self, page_id: str) -> list[dict]:
        """Fetch all immediate child pages of a given page.

        Returns list of dicts with keys: id, title, status.
        Handles cursor-based pagination automatically.
        """
        url = f"{self.base_url}/wiki/api/v2/pages/{page_id}/children"
        all_pages = []
        params = {"limit": 50}

        while True:
            data = self._get(url, params=params)
            all_pages.extend(data.get("results", []))

            next_link = data.get("_links", {}).get("next")
            if not next_link:
                break
            # next_link is a relative path; build full URL
            url = f"{self.base_url}{next_link}"
            params = None  # params are embedded in next_link

        return all_pages

    def get_release_feature_pages(self, production_parent_id: str) -> list[dict]:
        """Walk the Confluence hierarchy to find all customer-facing release pages.

        Expected hierarchy:
          Production Environment (production_parent_id)
            → Production-YYYY folders
              → COG Technical Release Notes(Production)-*
                → COG Release Features(Production)-*  ← primary target
              → COG Technical Patch Release Notes(Production)-*  ← also included

        For Technical Release Notes that have a Features child page, we
        return the Features child (that's the customer-facing content).
        For ones with no Features child, we return the Technical page itself
        (some years store content directly on it).
        Patch pages are leaf pages and are returned directly.

        Only pages matching known release-note naming patterns are included.
        Everything else is silently skipped.

        Returns list of dicts: {id, title, type} where type is
        "features" or "patch".
        """
        result = []

        # Level 1: Get Production-YYYY year folders
        year_folders = self.get_child_pages(production_parent_id)

        # Filter to only Production-YYYY / Production Releases-YYYY folders
        year_folders = [
            f for f in year_folders
            if f["title"].lower().startswith("production")
        ]
        logger.info("Found %d year folder(s) under Production Environment",
                     len(year_folders))

        for year_folder in year_folders:
            year_title = year_folder["title"]
            logger.info("  Scanning %s...", year_title)

            # Level 2: Get release note pages within the year folder
            release_pages = self.get_child_pages(year_folder["id"])

            for page in release_pages:
                title = page["title"]

                if "Patch" in title and "Release Notes" in title:
                    # Patch release notes are leaf pages (no Features child)
                    result.append({
                        "id": page["id"],
                        "title": title,
                        "type": "patch",
                    })
                    logger.info("    Found patch: %s", title)

                elif "Technical Release Notes" in title or "Technical_Release Notes" in title:
                    # Technical release notes may have a Features child page
                    children = self.get_child_pages(page["id"])
                    features_children = [
                        c for c in children
                        if "Features" in c["title"] or "Release Features" in c["title"]
                    ]

                    if features_children:
                        # Use the Features child page (customer-facing content)
                        for child in features_children:
                            result.append({
                                "id": child["id"],
                                "title": child["title"],
                                "type": "features",
                            })
                            logger.info("    Found features: %s", child["title"])
                    else:
                        # No Features child — content is on the Technical page itself
                        result.append({
                            "id": page["id"],
                            "title": title,
                            "type": "features",
                        })
                        logger.info("    Found features (no child): %s", title)

                else:
                    # Skip non-release-note pages silently
                    logger.debug("    Skipping non-release page: %s", title)

        return result

    def get_page_content(self, page_id: str) -> str:
        """Fetch page body in storage format (XHTML).

        Returns the raw XHTML string from Confluence storage format.
        """
        url = f"{self.base_url}/wiki/api/v2/pages/{page_id}"
        params = {"body-format": "storage"}
        data = self._get(url, params=params)
        return data.get("body", {}).get("storage", {}).get("value", "")

    def get_page_title(self, page_id: str) -> str:
        """Fetch just the page title."""
        url = f"{self.base_url}/wiki/api/v2/pages/{page_id}"
        data = self._get(url)
        return data.get("title", "")

    # ── Attachment operations (REST API v1 for downloads) ────────────

    def get_page_attachments(self, page_id: str) -> list[dict]:
        """List all attachments on a page.

        Returns list of dicts with keys: title (filename), mediaType,
        fileSize, _links.download.
        """
        url = (f"{self.base_url}/wiki/rest/api/content/{page_id}"
               f"/child/attachment")
        params = {"limit": 100, "expand": "version"}
        all_attachments = []

        while True:
            data = self._get(url, params=params)
            all_attachments.extend(data.get("results", []))

            next_link = data.get("_links", {}).get("next")
            if not next_link:
                break
            url = f"{self.base_url}{next_link}"
            params = None

        return all_attachments

    def download_attachment(self, download_path: str, dest: Path) -> None:
        """Download an attachment binary to a local file.

        Args:
            download_path: The _links.download value from the attachment.
            dest: Local Path to write the file to.
        """
        url = f"{self.base_url}/wiki{download_path}"
        resp = self.session.get(url, stream=True)
        resp.raise_for_status()

        dest.parent.mkdir(parents=True, exist_ok=True)
        with open(dest, "wb") as f:
            for chunk in resp.iter_content(chunk_size=8192):
                f.write(chunk)

File: ConfluenceToFlare/ConfluenceToFlare/Services/ConfluenceClient.swift
import Foundation
import os

/// Confluence Cloud REST API client.
///
/// Uses basic auth (email + API token) to fetch release note pages,
/// their content in storage format (XHTML), and image attachments.
/// Implements exponential backoff on 429 rate limit responses.
actor ConfluenceClient {
    private let baseURL: String
    private let session: URLSession
    private let logger = Logger(subsystem: "ConfluenceToFlare", category: "ConfluenceClient")

    init(baseURL: String, email: String, apiToken: String) {
        self.baseURL = baseURL.hasSuffix("/") ? String(baseURL.dropLast()) : baseURL

        let config = URLSessionConfiguration.default
        let credentials = "\(email):\(apiToken)"
        let base64 = Data(credentials.utf8).base64EncodedString()
        config.httpAdditionalHeaders = [
            "Authorization": "Basic \(base64)",
            "Accept": "application/json",
        ]
        self.session = URLSession(configuration: config)
    }

    // MARK: - Errors

    enum ClientError: LocalizedError {
        case rateLimitExceeded
        case httpError(Int)
        case invalidResponse
        case invalidURL(String)

        var errorDescription: String? {
            switch self {
            case .rateLimitExceeded:
                return "Rate limit exceeded after 5 retries"
            case .httpError(let code):
                return "HTTP error \(code)"
            case .invalidResponse:
                return "Invalid response from Confluence"
            case .invalidURL(let url):
                return "Invalid URL: \(url)"
            }
        }
    }

    // MARK: - Core Request with Retry

    /// GET request with exponential backoff on 429.
    private func get(url: String, params: [String: String]? = nil) async throws -> Data {
        guard var components = URLComponents(string: url) else {
            throw ClientError.invalidURL(url)
        }

        if let params {
            components.queryItems = params.map { URLQueryItem(name: $0.key, value: $0.value) }
        }

        guard let requestURL = components.url else {
            throw ClientError.invalidURL(url)
        }

        for attempt in 0..<5 {
            let (data, response) = try await session.data(from: requestURL)

            guard let httpResponse = response as? HTTPURLResponse else {
                throw ClientError.invalidResponse
            }

            if httpResponse.statusCode == 429 {
                let wait = UInt64(pow(2.0, Double(attempt))) * 1_000_000_000
                logger.warning("Rate limited (429), waiting \(Int(pow(2.0, Double(attempt))))s...")
                try await Task.sleep(nanoseconds: wait)
                continue
            }

            guard (200...299).contains(httpResponse.statusCode) else {
                throw ClientError.httpError(httpResponse.statusCode)
            }

            return data
        }

        throw ClientError.rateLimitExceeded
    }

    /// GET request that returns decoded JSON.
    private func getJSON<T: Decodable>(_ type: T.Type, url: String, params: [String: String]? = nil) async throws -> T {
        let data = try await get(url: url, params: params)
        return try JSONDecoder().decode(type, from: data)
    }

    // MARK: - Page Operations (REST API v2)

    /// Fetch all immediate child pages of a given page.
    /// Handles cursor-based pagination automatically.
    func getChildPages(pageID: String) async throws -> [PageResult] {
        var url = "\(baseURL)/wiki/api/v2/pages/\(pageID)/children"
        var params: [String: String]? = ["limit": "50"]
        var allPages: [PageResult] = []

        while true {
            let response = try await getJSON(PageListResponse.self, url: url, params: params)
            allPages.append(contentsOf: response.results)

            guard let nextLink = response._links?.next else { break }
            url = "\(baseURL)\(nextLink)"
            params = nil // params are embedded in next_link
        }

        return allPages
    }

    /// Walk the Confluence hierarchy to find all customer-facing release pages.
    ///
    /// Expected hierarchy:
    ///   Production Environment (productionParentID)
    ///     → Production-YYYY folders
    ///       → COG Technical Release Notes(Production)-*
    ///         → COG Release Features(Production)-*  ← primary target
    ///       → COG Technical Patch Release Notes(Production)-*  ← also included
    func getReleaseFeaturePages(productionParentID: String) async throws -> [(id: String, title: String, type: String)] {
        var result: [(id: String, title: String, type: String)] = []

        // Level 1: Get Production-YYYY year folders
        let yearFolders = try await getChildPages(pageID: productionParentID)
        let productionFolders = yearFolders.filter {
            $0.title.lowercased().hasPrefix("production")
        }
        logger.info("Found \(productionFolders.count) year folder(s) under Production Environment")

        for yearFolder in productionFolders {
            logger.info("  Scanning \(yearFolder.title)...")

            // Level 2: Get release note pages within the year folder
            let releasePages = try await getChildPages(pageID: yearFolder.id)

            for page in releasePages {
                let title = page.title

                if title.contains("Patch") && title.contains("Release Notes") {
                    result.append((id: page.id, title: title, type: "patch"))
                    logger.info("    Found patch: \(title)")

                } else if title.contains("Technical Release Notes") || title.contains("Technical_Release Notes") {
                    let children = try await getChildPages(pageID: page.id)
                    let featuresChildren = children.filter {
                        $0.title.contains("Features") || $0.title.contains("Release Features")
                    }

                    if !featuresChildren.isEmpty {
                        for child in featuresChildren {
                            result.append((id: child.id, title: child.title, type: "features"))
                            logger.info("    Found features: \(child.title)")
                        }
                    } else {
                        result.append((id: page.id, title: title, type: "features"))
                        logger.info("    Found features (no child): \(title)")
                    }
                } else {
                    logger.debug("    Skipping non-release page: \(title)")
                }
            }
        }

        return result
    }

    /// Fetch page body in storage format (XHTML).
    func getPageContent(pageID: String) async throws -> String {
        let url = "\(baseURL)/wiki/api/v2/pages/\(pageID)"
        let params = ["body-format": "storage"]
        let response = try await getJSON(PageResult.self, url: url, params: params)
        return response.body?.storage?.value ?? ""
    }

    /// Fetch just the page title.
    func getPageTitle(pageID: String) async throws -> String {
        let url = "\(baseURL)/wiki/api/v2/pages/\(pageID)"
        let response = try await getJSON(PageResult.self, url: url)
        return response.title
    }

    // MARK: - Attachment Operations (REST API v1)

    /// List all attachments on a page.
    func getPageAttachments(pageID: String) async throws -> [AttachmentResult] {
        var url = "\(baseURL)/wiki/rest/api/content/\(pageID)/child/attachment"
        var params: [String: String]? = ["limit": "100", "expand": "version"]
        var allAttachments: [AttachmentResult] = []

        while true {
            let response = try await getJSON(AttachmentListResponse.self, url: url, params: params)
            allAttachments.append(contentsOf: response.results)

            guard let nextLink = response._links?.next else { break }
            url = "\(baseURL)\(nextLink)"
            params = nil
        }

        return allAttachments
    }

    /// Download an attachment binary to a local file.
    func downloadAttachment(downloadPath: String, dest: URL) async throws {
        let url = "\(baseURL)/wiki\(downloadPath)"
        guard let requestURL = URL(string: url) else {
            throw ClientError.invalidURL(url)
        }

        let (tempURL, response) = try await session.download(from: requestURL)

        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw ClientError.httpError((response as? HTTPURLResponse)?.statusCode ?? 0)
        }

        let fm = FileManager.default
        try fm.createDirectory(at: dest.deletingLastPathComponent(), withIntermediateDirectories: true)

        if fm.fileExists(atPath: dest.path) {
            try fm.removeItem(at: dest)
        }
        try fm.moveItem(at: tempURL, to: dest)
    }
}
